File: leetcode/2807.在链表中插入最大公约数.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # 直接遍历
        arr = []
        cur = head
        while cur:
            arr.append(cur.val)
            cur = cur.next
        print(arr)
        n = len(arr)
        if n == 0 or n == 1: return head
        ans = []
        for i in range(1, n):
            ans.append(arr[i-1])
            ans.append(gcd(arr[i-1], arr[i]))
        ans.append(arr[-1])
        print(ans)
        root = ListNode(0)
        cur = root
        for x in ans:
            Node = ListNode(x)
            cur.next = Node
            cur = cur.next
        return root.next
            
            
            

File: leetcode/2807.在链表中插入最大公约数.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertGreatestCommonDivisors(head *ListNode) *ListNode {
    node := head
    for node.Next != nil {
        node.Next = &ListNode{gcd(node.Val, node.Next.Val), node.Next}
        node = node.Next.Next
    }
    return head
}

func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a % b
    }
    return a
}